pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -1064,3 +1064,87 @@ func setUpstreamProtocol(node *model.Proxy, c *cluster.Cluster, port *model.Port
 		c.ProtocolSelection = cluster.Cluster_USE_DOWNSTREAM_PROTOCOL
 	}
 }
+
+func addTelemetryMetadata(opts buildClusterOpts, service *model.Service, direction model.TrafficDirection) {
+	if opts.cluster == nil {
+		return
+	}
+	if direction == model.TrafficDirectionInbound && (opts.proxy == nil || opts.proxy.ServiceInstances == nil ||
+		len(opts.proxy.ServiceInstances) == 0 || opts.port == nil) {
+		// At inbound, port and local service instance has to be provided
+		return
+	}
+	if direction == model.TrafficDirectionOutbound && service == nil {
+		// At outbound, the service corresponding to the cluster has to be provided.
+		return
+	}
+	if opts.cluster.Metadata == nil {
+		opts.cluster.Metadata = &core.Metadata{
+			FilterMetadata: map[string]*structpb.Struct{},
+		}
+	}
+
+	// Create Istio metadata if does not exist yet
+	if _, ok := opts.cluster.Metadata.FilterMetadata[util.IstioMetadataKey]; !ok {
+		opts.cluster.Metadata.FilterMetadata[util.IstioMetadataKey] = &structpb.Struct{
+			Fields: map[string]*structpb.Value{},
+		}
+	}
+
+	im := opts.cluster.Metadata.FilterMetadata[util.IstioMetadataKey]
+
+	// Add services field into istio metadata
+	im.Fields["services"] = &structpb.Value{
+		Kind: &structpb.Value_ListValue{
+			ListValue: &structpb.ListValue{
+				Values: []*structpb.Value{},
+			},
+		},
+	}
+
+	svcMetaList := im.Fields["services"].GetListValue()
+
+	// Add servie related metadata. This will be consumed by telemetry v2 filter for metric labels.
+	if direction == model.TrafficDirectionInbound {
+		// For inbound cluster, add all services on the cluster port
+		for _, svc := range opts.proxy.ServiceInstances {
```

Comment:
There can be potentionally 100s of duplicates because ServiceInstances may have multiple per IP. On my cluster, I get:
```

       "filter_metadata": {
        "istio": {
         "services": [
          {
           "namespace": "default",
           "host": "shell.default.svc.cluster.local",
           "name": "shell"
          },
          {
           "name": "shell",
           "namespace": "default",
           "host": "shell.default.svc.cluster.local"
          },
          {
           "namespace": "default",
           "host": "shell.default.svc.cluster.local",
           "name": "shell"
          },
          {
           "host": "shell.default.svc.cluster.local",
           "name": "shell",
           "namespace": "default"
          },
          {
           "host": "shell.default.svc.cluster.local",
           "name": "shell",
           "namespace": "default"
          },
          {
           "host": "shell.default.svc.cluster.local",
           "name": "shell",
           "namespace": "default"
          },
          {
           "name": "shell",
           "namespace": "default",
           "host": "shell.default.svc.cluster.local"
          },
          {
           "namespace": "default",
           "host": "shell.default.svc.cluster.local",
           "name": "shell"
          },
          {
           "namespace": "default",
           "host": "shell.default.svc.cluster.local",
           "name": "shell"
          },
          {
           "host": "shell.default.svc.cluster.local",
           "name": "shell",
           "namespace": "default"
          },
          {
           "namespace": "default",
           "host": "shell.default.svc.cluster.local",
           "name": "shell"
          },
          {
           "name": "shell",
           "namespace": "default",
           "host": "shell.default.svc.cluster.local"
          },
          {
           "name": "shell",
           "namespace": "default",
           "host": "shell.default.svc.cluster.local"
          },
          {
           "namespace": "default",
           "host": "shell.default.svc.cluster.local",
           "name": "shell"
          }
         ]
        }
       }
      },
```

Reply:
maybe need the same logic here - https://github.com/istio/istio/blob/3b1d29a8abcbff5af5d3dab01103e1736b3f8f31/pilot/pkg/networking/core/v1alpha3/cluster.go#L288-L296

Reply:
So they have different protocol or name for the service port? Here I only dedup based on port number, so instead I should dedup them based on the whole port struct?

Reply:
Actually, sorry I think I made a mistake on the dedup condition. Maybe that's why it is causing the issue.